Include/internal/pycore_critical_section.h

```diff
@@ -0,0 +1,204 @@
+#ifndef Py_INTERNAL_CRITICAL_SECTION_H
+#define Py_INTERNAL_CRITICAL_SECTION_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_lock.h"        // PyMutex
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Implementation of Python critical sections: helpers to replace the global
+// interpreter lock with fine grained locking.
+//
+// A Python critical section is a region of code that can only be executed by
+// a single thread at at time. The regions begin with a call to
+// Py_BEGIN_CRITICAL_SECTION and ends with Py_END_CRITICAL_SECTION. Active
+// critical sections are *implicitly* suspended whenever the thread calls
+// `_PyThreadState_Detach()` (i.e., at tiems when the interpreter would have
+// released the GIL).
+//
+// The most recent critical section is resumed when `_PyThreadState_Attach()`
+// is called. See `_PyCriticalSection_Resume()`.
+//
+// The purpose of implicitly ending critical sections is to avoid deadlocks
+// when locking multiple objects. Any time a thread would have released the
+// GIL, it releases all locks from active critical sections. This includes
+// times when a thread blocks while attempting to acquire a lock.
+//
+// Note that the macros Py_BEGIN_CRITICAL_SECTION and Py_END_CRITICAL_SECTION
+// are no-ops in non-free-threaded builds.
+//
+// Example usage:
+//  Py_BEGIN_CRITICAL_SECTION(op);
+//  ...
+//  Py_END_CRITICAL_SECTION;
+//
+// To lock two objects at once:
+//  Py_BEGIN_CRITICAL_SECTION2(op1, op2);
+//  ...
+//  Py_END_CRITICAL_SECTION2;
```

Comment:
No, it's not the same because the inner critical sections might suspend the outer one. You In your first example, you are only guaranteed to hold the lock for `op2` within the `...`. In the second example, you are only guaranteed to hold the lock for `op5` within the `...`.

I've added a NOTE about this.

Reply:
Thanks for clarifying.  That behavior with explicitly nested sections is both counter to my intuition and expected (else you wouldn't have bothered with a dedicated 2-lock API).

I suppose that isn't such a big deal since this is an internal-only API and we will adjust to our needs when necessary without much fanfare. 😄